File: ios/Classes/FlutterTwilioError.swift
import Foundation
import TwilioVoice

/// Single source of truth for wire-format errors sent to Dart via Pigeon.
///
/// Throw the returned [PigeonError] from a VoiceHostApi method; Pigeon's
/// generated `wrapError` helper serializes it as a Dart `PlatformException`
/// with a stable `code`, matching the Android `FlutterError` Kotlin class.
///
/// (Swift's `FlutterError` is Objective-C and doesn't conform to `Error`, so
/// we tunnel all typed errors through `PigeonError` which does.)
enum FlutterTwilioError {

    // MARK: - Twilio error-code groupings
    //
    // These ranges are centralised here so TVCallDelegate, TVRegistrationHandler
    // and the call-continuation logic all agree on what counts as "invalid
    // token" vs. "connection error" vs. "twilio_sdk_error".

    /// Twilio access-token related failures: invalid token, expired, malformed,
    /// wrong audience, rejected by auth service.
    /// See `TVOError.h`: 20101–20107, 20151, 20157, 51007.
    static let tokenErrorCodes: Set<Int> = [
        20101, 20102, 20103, 20104, 20105, 20106, 20107,
        20151, 20157, 51007,
    ]

    /// Transport / network / signaling level failures. The call got as far as
    /// the SDK but the underlying socket, DNS or media plane gave up.
    /// Covers 31005 (connection), 31009 (transport), 31530 (DNS),
    /// 53001 (signalling), 53405 (media connection).
    static let connectionErrorCodes: Set<Int> = [
        31005, 31009, 31530, 53001, 53405,
    ]

    static func of(_ code: String, _ message: String, _ details: [String: Any?] = [:]) -> PigeonError {
        PigeonError(code: code, message: message, details: details)
    }

    /// Maps a Twilio SDK error to a [PigeonError]. Returns an `invalid_token`
    /// or `connection_error` when the NSError code lands in the corresponding
    /// set, else falls through to `twilio_sdk_error`.
    static func fromTwilio(_ err: Error) -> PigeonError {
        let ns = err as NSError
        let twilioCode = ns.userInfo["TVErrorCodeKey"] as? Int ?? ns.code
        let code = stableCodeFor(nsError: ns, twilioCode: twilioCode)
        return PigeonError(
            code: code,
            message: ns.localizedDescription,
            details: [
                "twilioCode": twilioCode,
                "twilioDomain": ns.domain,
                "nativeMessage": ns.localizedDescription,
            ]
        )
    }

    /// Like [fromTwilio] but hard-codes the `code` to `twilio_sdk_error` — used
    /// in the rare case where we want to bypass the stable-code mapping.
    static func rawTwilio(_ err: Error) -> PigeonError {
        let ns = err as NSError
        let twilioCode = ns.userInfo["TVErrorCodeKey"] as? Int ?? ns.code
        return PigeonError(
            code: "twilio_sdk_error",
            message: ns.localizedDescription,
            details: [
                "twilioCode": twilioCode,
                "twilioDomain": ns.domain,
                "nativeMessage": ns.localizedDescription,
            ]
        )
    }

    static func unknown(_ err: Error) -> PigeonError {
        let ns = err as NSError
        return PigeonError(
            code: "unknown",
            message: ns.localizedDescription,
            details: [
                "nativeMessage": ns.localizedDescription,
                "cause": String(describing: type(of: err)),
            ]
        )
    }

    /// Shared classifier used by the Call.Delegate, RegistrationHandler and
    /// `fromTwilio(_:)` so every surface emits the same stable code for the
    /// same NSError.
    static func stableCodeFor(nsError ns: NSError, twilioCode: Int) -> String {
        if tokenErrorCodes.contains(twilioCode) {
            return "invalid_token"
        }
        if connectionErrorCodes.contains(twilioCode) {
            return "connection_error"
        }
        if ns.domain == NSURLErrorDomain {
            return "connection_error"
        }
        return "twilio_sdk_error"
    }

    /// Details map commonly attached to error events so Dart can surface the
    /// underlying Twilio diagnostics.
    static func twilioDetails(_ err: Error) -> [String: Any?] {
        let ns = err as NSError
        let twilioCode = ns.userInfo["TVErrorCodeKey"] as? Int ?? ns.code
        return [
            "twilioCode": twilioCode,
            "twilioDomain": ns.domain,
            "nativeMessage": ns.localizedDescription,
        ]
    }
}

File: android/src/main/kotlin/com/dev/flutter_twilio/FlutterTwilioError.kt
package com.dev.flutter_twilio

import com.dev.flutter_twilio.generated.FlutterError
import com.twilio.voice.CallException
import com.twilio.voice.RegistrationException

/**
 * Single source of truth for wire-format errors sent to Dart via Pigeon.
 * Return / throw the produced [FlutterError] from a HostApi method — it lands
 * as a PlatformException on the Dart side with a stable [code].
 */
object FlutterTwilioError {

    /**
     * Twilio access-token error codes. Per Twilio Voice SDK docs these cover
     * invalid, expired, or otherwise unusable JWT credentials — we surface them
     * as [`invalid_token`] instead of the generic [`twilio_sdk_error`].
     */
    val TOKEN_ERROR_CODES: Set<Int> = setOf(20101, 20104, 20157)

    /**
     * Twilio signaling error range (53xxx). Covers transport / connection
     * failures we surface as [`connection_error`].
     */
    fun isSignalingError(code: Int): Boolean = code in 53000..53999

    fun of(
        code: String,
        message: String,
        details: Map<String, Any?> = emptyMap(),
    ): FlutterError = FlutterError(code, message, details)

    fun fromTwilio(sdk: CallException): FlutterError {
        val twilioCode = sdk.errorCode
        val (code, message) = when {
            twilioCode in TOKEN_ERROR_CODES -> "invalid_token" to (sdk.message ?: "Access token invalid")
            isSignalingError(twilioCode) -> "connection_error" to (sdk.message ?: "Signaling error")
            else -> "twilio_sdk_error" to (sdk.message ?: "Twilio SDK error")
        }
        return of(
            code = code,
            message = message,
            details = mapOf(
                "twilioCode" to twilioCode,
                "twilioDomain" to "com.twilio.voice",
                "nativeMessage" to sdk.message,
            ),
        )
    }

    fun fromRegistration(sdk: RegistrationException): FlutterError {
        val twilioCode = sdk.errorCode
        val code = if (twilioCode in TOKEN_ERROR_CODES) "invalid_token" else "registration_error"
        return of(
            code = code,
            message = sdk.message ?: "Registration failure",
            details = mapOf(
                "twilioCode" to twilioCode,
                "twilioDomain" to "com.twilio.voice",
                "nativeMessage" to sdk.message,
            ),
        )
    }

    fun bluetoothUnavailable(message: String): FlutterError =
        of("bluetooth_unavailable", message)

    fun wiredUnavailable(message: String): FlutterError =
        of("wired_unavailable", message)

    fun audioRouteFailed(message: String): FlutterError =
        of("audio_route_failed", message)

    fun toneAssetNotFound(path: String): FlutterError =
        of(
            code = "tone_asset_not_found",
            message = "Tone asset not found: $path",
            details = mapOf("assetPath" to path),
        )

    fun notificationPermissionDenied(): FlutterError =
        of(
            code = "notification_permission_denied",
            message = "POST_NOTIFICATIONS permission is required",
        )

    fun unknown(cause: Throwable): FlutterError = of(
        code = "unknown",
        message = cause.message ?: cause::class.java.simpleName,
        details = mapOf(
            "nativeMessage" to cause.message,
            "cause" to cause::class.java.name,
        ),
    )
}
